### PR TITLE
APIGOV-22657 Sample subscription

### PR DESCRIPTION
**Pull request by tsjohns9**

### CODE REVIEW

File: pkg/traceability/sampling/sample.go [hunk 1]
```diff
@@ -20,9 +20,28 @@ func (s *sample) ShouldSampleTransaction(details TransactionDetails) bool {
...
-	if s.config.PerAPI && details.APIID != "" {
-		return s.shouldSampleWithCounter(details.APIID)
+
+	if s.config.PerAPI && s.config.PerSub {
+		if details.SubID != "" {
+			return s.shouldSampleWithCounter(details.SubID)
+		}
+		if details.APIID != "" {
+			return s.shouldSampleWithCounter(details.APIID)
+		}
```
Comment (jcollins-axway) [inline pkg/traceability/sampling/sample.go:30]: I think we should also have a concatenation of the subid and apiid

File: pkg/traceability/sampling/sample.go [hunk 2]
```diff
@@ -20,9 +21,28 @@ func (s *sample) ShouldSampleTransaction(details TransactionDetails) bool {
 	if hasFailedStatus && s.config.ReportAllErrors {
 		return true
 	}
-	if s.config.PerAPI && details.APIID != "" {
-		return s.shouldSampleWithCounter(details.APIID)
+
+	if s.config.PerAPI && s.config.PerSub {
+		if details.SubID != "" {
+			return s.shouldSampleWithCounter(fmt.Sprintf("%s-%s", details.APIID, details.SubID))
```
Comment (tsjohns9) [inline pkg/traceability/sampling/sample.go:27]: @jcollins-axway  like this?

Comment (jcollins-axway) [reply pkg/traceability/sampling/sample.go:27]: that looks better, think you can drop the conditional on 29 and just return then

Comment (jcollins-axway) [reply pkg/traceability/sampling/sample.go:27]: that looks better, think you can drop the conditional on 29 and just return then

Comment (tsjohns9) [reply pkg/traceability/sampling/sample.go:27]: We still need to check if there is an api id though?

Comment (jcollins-axway) [reply pkg/traceability/sampling/sample.go:27]: true true, that works

File: pkg/traceability/sampling/globalsampling.go [hunk 3]
```diff
@@ -14,6 +14,7 @@ var agentSamples *sample
 type Sampling struct {
 	Percentage      int  `config:"percentage"    validate:"min=0, max=100"`
 	PerAPI          bool `config:"per_api"`
+	PerSub          bool `config:"per_api"`
```
Comment (jcollins-axway) [inline pkg/traceability/sampling/globalsampling.go:17]: config setting needs updated

Comment (tsjohns9) [reply pkg/traceability/sampling/globalsampling.go:17]: What about it needs to be updated?

File: pkg/traceability/sampling/sampling_test.go [hunk 4]
```diff
@@ -192,6 +193,64 @@ func TestShouldSample(t *testing.T) {
...
+				"id3": 50, // expect 50
+				"id4": 50, // expect 50
+			},
+			subIDs: map[string]string{
+				"id1": "sub1",
+				"id2": "sub2",
+				"id3": "sub3",
+				"id4": "sub4",
+			},
+			expectedSampled: 200,
```
Comment (tsjohns9) [inline pkg/traceability/sampling/sampling_test.go:210]: Why is the expectedSampled at 200 when the percent is 50 and the total requests is 200? I copied this from another test which is set up the same way.

File: pkg/traceability/sampling/globalsampling.go (hunk 3 above)
Comment (jcollins-axway) [reply pkg/traceability/sampling/globalsampling.go:17]: PerSub bool `config:"per_api"`. -> PerSub bool `config:"per_subscription"`

File: pkg/traceability/sampling/sampling_test.go (hunk 4 above)
Comment (jcollins-axway) [reply pkg/traceability/sampling/sampling_test.go:210]: 4 ids and we are sampling by id. So each of them will be guaranteed 50 of every 100 transactions

Comment (jcollins-axway) [reply pkg/traceability/sampling/sampling_test.go:210]: The way our sampling works, with the counters, the first 50 of each id will be sampled, the next 50 of each id will not be.  So if we then set all ids to send 75 transactions (300 total) we would still expect 200 sampled.

File: pkg/traceability/sampling/sample.go [hunk 5]
```diff
@@ -20,9 +21,28 @@ func (s *sample) ShouldSampleTransaction(details TransactionDetails) bool {
...
+		}
+		if details.APIID != "" {
+			return s.shouldSampleWithCounter(details.APIID)
+		}
+	}
```
Comment (jcollins-axway) [inline pkg/traceability/sampling/sample.go:32]: I was thinking, in this case where both settings are on.

We probably want to count every apiid, even if subid and apiid are both set.  That way when subid is not set we still have the counts of all transactions for that apiid.

Comment (tsjohns9) [reply pkg/traceability/sampling/sample.go:32]: Ok, what should change here?

Comment (jcollins-axway) [reply pkg/traceability/sampling/sample.go:32]: Something like this, maybe?

```
if s.config.PerAPI && s.config.PerSub {
	if details.SubID != "" {
		s.shouldSampleWithCounter(details.APIID). // keep count of sampling by just api id as well
		return s.shouldSampleWithCounter(fmt.Sprintf("%s-%s", details.APIID, details.SubID))
	}
	if details.APIID != "" {
		return s.shouldSampleWithCounter(details.APIID)
	}
}
```

**Review by jcollins-axway — APPROVED** [review]

**Pull request merged**